### PR TITLE
Advanced network configuration for Puppet

### PR DESCRIPTION
**Pull request by xenolog**

### CODE REVIEW

**Comment by Zipfer** [comment]
change all "if not"
use !,&&,|| instead use not,and,or 
https://github.com/bbatsov/ruby-style-guide

I'm did not edit the entry code, I only showed a few examples

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+    # define internal puppet parameters for creating resources
+    res_factory = {
+      :br      => { :name_of_resource => 'l23network::l2::bridge' },
+      :port    => { :name_of_resource => 'l23network::l2::port' },
+      :bond    => { :name_of_resource => 'l23network::l2::bond' },
+      :patch   => { :name_of_resource => 'l23network::l2::patch' },
+      :ifconfig=> { :name_of_resource => 'l23network::l3::ifconfig' }
+    }
+    res_factory.each do |k, v|
```
Comment (aglarendil) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:147]: comment this line please. what '::' means?

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+      else
+        # operate by custom Type
+        res_factory[k][:resource] = Puppet::Type.type(v[:name_of_resource].to_sym())
+        res_factory[k][:type_of_resource] = :type
+      end
+    end
+
+    # collect interfaces and endpoints
+    endpoints = {}
+    borned_ports = []
```
Comment (aglarendil) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:161]: s/borned/born/

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    config_hash[:interfaces].each do |int_name, int_properties|
+      int_name = int_name.to_sym()
+      endpoints[int_name] = create_endpoint()
+      borned_ports.insert(-1, int_name)
```
Comment (aglarendil) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:165]: s/borned/born/

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    config_hash[:endpoints].each do |e_name, e_properties|
+      e_name = e_name.to_sym()
+      if not endpoints[e_name]
+        endpoints[e_name] = create_endpoint()
+      end
+      e_properties.each do |k,v|
+        if k.to_sym() == :IP
+          if not (v.is_a?(Array) or ['none','dhcp'].index(v))
+            raise(Puppet::ParseError, "IP field for endpoint '#{e_name}' must be array of IP addresses, 'dhcp' or 'none'.")
+          elsif ['none','dhcp'].index(v)
```
Comment (aglarendil) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:176]: may be we should use "include?"

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+            end
+          end
+        else
+          endpoints[e_name][:properties][k.to_sym()] = v
+        end
+      end
+    end
+
+    # execute transformations
+    # todo: if pribider="lnx" execute transformations for LNX bridges
```
Comment (aglarendil) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:195]: typos

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    transformation_success = []
+    previous = nil
+    config_hash[:transformations].each do |t|
+      action = t[:action].strip()
+      if action.start_with?('add-')
```
Comment (aglarendil) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:200]: may be we should update specification to have two different fields, e.g. "action: add|delete|update" and "type: bond|bridge|vlan|..."

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+      trans = sanitize_transformation(t)
+      resource = res_factory[action][:resource]
+      p_resource = Puppet::Parser::Resource.new(
+          res_factory[action][:name_of_resource],
+          trans[:name],
+          :scope => self,
+          :source => resource
+      )
+      trans.select{|k,v| k != :action}.each do |k,v|
```
Comment (aglarendil) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:214]: I suggest to use separate hash of parameters and iterate through it.

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb (hunk 2 above)
Comment (xenolog) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:161]: fix

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb (hunk 3 above)
Comment (xenolog) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:165]: fix

File: deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb (hunk 5 above)
Comment (xenolog) [inline deployment/puppet/l23network/lib/puppet/parser/functions/config_network_from_json_v1.rb:195]: fix

**Comment by xenolog** [comment]
Will be applied later.

**Pull request closed without merging**